direct/tpm2/reflect.go

```diff
@@ -0,0 +1,1002 @@
+					v.Type().Field(i).Name, v.Type().Name())
+			}
+		}
+	}
+	return nil
+}
+
+// Unmarshals a bitwise-defined struct.
+func unmarshalBitwise(buf *bytes.Buffer, v reflect.Value) error {
+	maxBit := 0
+	for i := 0; i < v.NumField(); i++ {
+		high, _, ok := rangeTag(v.Type().Field(i), "bit")
+		if !ok {
+			return fmt.Errorf("'%v' struct member '%v' did not specify a bit index or range", v.Type().Name(), v.Type().Field(i).Name)
+		}
+		if high > maxBit {
+			maxBit = high
+		}
+	}
+	if (maxBit+1)%8 != 0 {
+		return fmt.Errorf("'%v' bitwise members did not total up to a multiple of 8 bits", v.Type().Name())
+	}
+	bitArray := make([]bool, maxBit+1)
+	// We will read big-endian, starting from the last byte and working our
+	// way down.
+	for i := len(bitArray)/8 - 1; i >= 0; i-- {
+		b, err := buf.ReadByte()
+		if err != nil {
+			return fmt.Errorf("error %d bits into field '%v' of struct '%v': %w",
+				i, v.Type().Field(i).Name, v.Type().Name(), err)
+		}
+		for j := 0; j < 8; j++ {
+			bitArray[8*i+j] = (((b >> j) & 1) == 1)
+		}
+	}
+	for i := 0; i < v.NumField(); i++ {
+		high, low, _ := rangeTag(v.Type().Field(i), "bit")
+		var val uint64
+		for j := 0; j <= high-low; j++ {
+			if bitArray[low+j] {
+				val |= (1 << j)
+			}
+		}
+		if v.Field(i).Kind() == reflect.Bool {
+			v.Field(i).SetBool((val & 1) == 1)
+		} else {
+			v.Field(i).SetUint(val)
+		}
+	}
+	return nil
+}
+
+// Unmarshals the member of the given union struct corresponding to the given
+// selector. Unmarshals nothing if the selector is TPM_ALG_NULL (0x0010).
+func unmarshalUnion(buf *bytes.Buffer, v reflect.Value, selector int64) error {
+	// Special case: TPM_ALG_NULL as a selector means unmarshal nothing
+	if selector == int64(tpm.AlgNull) {
+		return nil
+	}
+	for i := 0; i < v.NumField(); i++ {
+		sel, ok := numericTag(v.Type().Field(i), "selector")
+		if !ok {
+			return fmt.Errorf("'%v' union member '%v' did not have a selector tag", v.Type().Name(), v.Type().Field(i).Name)
+		}
+		if sel == selector {
+			val := reflect.New(v.Type().Field(i).Type.Elem())
+			if err := unmarshal(buf, val.Elem()); err != nil {
+				return fmt.Errorf("unmarshalling '%v' union member '%v': %w", v.Type().Name(), v.Type().Field(i).Name, err)
+			}
+			v.Field(i).Set(val)
+			return nil
+		}
+	}
+	return fmt.Errorf("selector value '%v' not handled for type '%v'", selector, v.Type().Name())
+}
+
+// Returns all the gotpm tags on a field as a map.
+// Some tags are settable (with "="). For these, the value is the RHS.
+// For all others, the value is the empty string.
+func tags(t reflect.StructField) map[string]string {
+	allTags, ok := t.Tag.Lookup("gotpm")
+	if !ok {
+		return nil
+	}
+	result := make(map[string]string)
+	tags := strings.Split(allTags, ",")
+	for _, tag := range tags {
+		// Split on the equals sign for settable tags.
+		// If the split returns an empty slice, this is an empty tag.
```

Comment:
Is this actually true? I thought we still get `len(assignment) == 1` if `tag == ""`

Reply:
Impressively spotted, fixed. Thanks!